File: benchmarks/strsv/wgblas/strsv.js
import { init, cleanup } from "wgblas";
import { strsv } from "wgblas/strsv";
import { GpuVector } from "wgblas/classes/GpuVector";
import { GpuMatrix } from "wgblas/classes/GpuMatrix";
import { randomFloat32Array, randomTriangularFloat32Array } from "wgblas/random";
import {
  median,
  printHeader,
  printRow,
  getGpuModel,
  saveResults,
  toColumnMajor,
} from "../../utils/helpers.mjs";

const WARMUP_ITERS = 5;
const BENCH_ITERS = 100;
// Square n×n triangular matrix; lower triangle stored. Blocked solve: each
// 64-row diagonal block is still a genuine sequential dependency (single
// workgroup, like strmv is NOT), but propagating a solved block onto the
// remaining rows is a dense, fully parallel update (same shape as strmv) —
// see strsv_block.wgsl/strsv_update.wgsl/strsv.mjs. That turns n sequential
// stages into ceil(n/64), so this scales to the same sizes strmv does.
//
// Unlike every other Level 2 routine here, strsv's dominant performance
// story isn't uplo/trans/lda — it's dispatch-count overhead: each block is
// ~2 sequential GPU dispatches (apply + update), so total time is bound by
// fixed per-dispatch overhead at small n and by growing per-block update
// work at large n. numBlocks and ms_per_block below make that curve
// directly visible instead of requiring the reader to derive it. uplo was
// scoped and found to be a minor effect (~2-4%, matches strmv); trans is
// real but modest (~1.2-1.3x, diluted since it only affects the "update"
// pass, a fraction of total dispatch-dominated time); lda (no-transpose)
// and diag are both confirmed non-effects — none of the four are swept as
// separate files, since they're not the story here.
const SIZES = [32, 64, 128, 256, 512, 1024, 2048, 4096];
const BLOCK_SIZE = 64; // matches strsv.mjs's own BLOCK_SIZE

const COLS = ["n", "numBlocks", "compute_ms", "ms_per_block", "compute_GBs"];

const powerPreference =
  process.argv[2] === "low-power" ? "low-power" : "high-performance";
const device = await init({ benchmark: true, powerPreference });

const gpuModel = getGpuModel();
const records = [];

printHeader(COLS);

for (const size of SIZES) {
  const n = size;
  const lda = n;

  // Well-conditioned diagonal (away from 0) — strsv divides by it. Generated
  // row-major then transposed to genuine column-major storage of the same
  // logical (still well-conditioned, still lower-triangular) matrix.
  const AGpu = GpuMatrix.from(
    toColumnMajor(randomTriangularFloat32Array(n, lda, "lower"), n, n), n, n, lda, "column-major",
  );
  const b = randomFloat32Array(n);

  // strsv solves in place, so reset x to b before every call, or later
  // iterations solve using drifted-toward-0 leftovers from earlier ones.
  // GpuVector has no COPY_DST usage, so writeBuffer can't target it —
  // destroy+recreate instead, before the timed strsv call.
  let xGpu = null;
  const resetX = () => {
    if (xGpu) xGpu.destroy();
    xGpu = GpuVector.from(b);
  };

  // warm up
  for (let i = 0; i < WARMUP_ITERS; i++) {
    resetX();
    await strsv(device, "lower", "no-transpose", "non-unit", n, AGpu, lda, xGpu, 1);
  }

  const times = [];
  for (let i = 0; i < BENCH_ITERS; i++) {
    resetX();
    const { gpuTimeMs } = await strsv(
      device, "lower", "no-transpose", "non-unit", n, AGpu, lda, xGpu, 1,
    );
    if (Number.isFinite(gpuTimeMs) && gpuTimeMs > 0) times.push(gpuTimeMs);
  }

  AGpu.destroy();
  xGpu.destroy();

  if (times.length === 0) continue;

  const med = median(times);
  const numBlocks = Math.ceil(n / BLOCK_SIZE);
  const msPerBlock = med / numBlocks;
  // lower triangle A read + x read + x write (in place)
  const bytes = (n * (n + 1) / 2 + n + n) * 4;
  const gbs = bytes / 1e9 / (med / 1e3);
  printRow(COLS, [n, numBlocks, med, msPerBlock, gbs]);
  records.push({ n, numBlocks, compute_ms: med, ms_per_block: msPerBlock, compute_GBs: gbs });
}

saveResults("strsv", gpuModel, records, { folder: "strsv" });

cleanup();

File: benchmarks/strsv/cuda/strsv.c
// Unlike every other Level 2 routine here, strsv's dominant performance
// story (in the wgblas implementation) isn't uplo/trans/lda — it's
// dispatch-count overhead from its blocked/sequential solve (see
// strsv.js). cuBLAS's cublasStrsv is an opaque single call with no
// equivalent block-dispatch decomposition to characterize the same way,
// so this stays a plain n sweep — no numBlocks/ms_per_block columns here.

#include <stdio.h>
#include <stdlib.h>
#include "../../utils/helpers.h"

#define WARMUP_ITERS 5
#define BENCH_ITERS  100

int main(void) {
    char gpu_model[256];
    get_gpu_model(gpu_model, sizeof(gpu_model));

    cublasHandle_t handle;
    CUBLAS_CHECK(cublasCreate(&handle));

    int sizes[] = { 32, 64, 128, 256, 512, 1024, 2048, 4096 };
    int num_sizes = (int)(sizeof(sizes) / sizeof(sizes[0]));

    float med_times[num_sizes];
    float gbs_vals[num_sizes];

    printf("%-10s  %-12s  %-12s\n", "n", "compute_ms", "compute_GBs");
    printf("%-10s  %-12s  %-12s\n", "----------", "------------", "------------");

    for (int si = 0; si < num_sizes; si++) {
        int n = sizes[si];
        int lda = n;

        float *h_A = random_float_array(n * n, -1.0f, 1.0f);
        float *h_x = random_float_array(n, -1.0f, 1.0f);

        float *d_A, *d_x;
        CUDA_CHECK(cudaMalloc((void **)&d_A, (size_t)n * n * sizeof(float)));
        CUDA_CHECK(cudaMalloc((void **)&d_x, n * sizeof(float)));
        CUDA_CHECK(cudaMemcpy(d_A, h_A, (size_t)n * n * sizeof(float), cudaMemcpyHostToDevice));

        cudaEvent_t start, stop;
        CUDA_CHECK(cudaEventCreate(&start));
        CUDA_CHECK(cudaEventCreate(&stop));

        // cublasStrsv solves op(A)*x = b in place (b on input, x on output).
        // d_A is genuinely column-major — cuBLAS's native layout, so
        // FILL_MODE_LOWER + OP_N read the same op(A) = A wgblas's
        // "lower"/"no-transpose" reads, no reinterpretation trick needed.
        for (int i = 0; i < WARMUP_ITERS; i++) {
            CUDA_CHECK(cudaMemcpy(d_x, h_x, n * sizeof(float), cudaMemcpyHostToDevice));
            CUBLAS_CHECK(cublasStrsv(handle, CUBLAS_FILL_MODE_LOWER, CUBLAS_OP_N, CUBLAS_DIAG_NON_UNIT, n, d_A, lda, d_x, 1));
        }
        CUDA_CHECK(cudaDeviceSynchronize());

        float compute_times[BENCH_ITERS];
        for (int i = 0; i < BENCH_ITERS; i++) {
            CUDA_CHECK(cudaMemcpy(d_x, h_x, n * sizeof(float), cudaMemcpyHostToDevice));
            CUDA_CHECK(cudaEventRecord(start, 0));
            CUBLAS_CHECK(cublasStrsv(handle, CUBLAS_FILL_MODE_LOWER, CUBLAS_OP_N, CUBLAS_DIAG_NON_UNIT, n, d_A, lda, d_x, 1));
            CUDA_CHECK(cudaEventRecord(stop, 0));
            CUDA_CHECK(cudaEventSynchronize(stop));
            CUDA_CHECK(cudaEventElapsedTime(&compute_times[i], start, stop));
        }

        float med = median(compute_times, BENCH_ITERS);
        // lower triangle A read + x read + x write (in place)
        float bytes = ((float)(n * (n + 1) / 2) + 2.0f * n) * 4.0f;
        float gbs = (bytes / 1e9f) / (med / 1e3f);

        med_times[si] = med;
        gbs_vals[si]  = gbs;

        printf("%-10d  %-12.4f  %-12.4f\n", n, med, gbs);

        CUDA_CHECK(cudaEventDestroy(start));
        CUDA_CHECK(cudaEventDestroy(stop));
        CUDA_CHECK(cudaFree(d_A));
        CUDA_CHECK(cudaFree(d_x));
        free(h_A);
        free(h_x);
    }

    save_results_ex("strsv", gpu_model, "strsv", "strsv", sizes, med_times, gbs_vals, num_sizes);

    CUBLAS_CHECK(cublasDestroy(handle));
    return 0;
}
